lib/customValidators.js

```diff
@@ -0,0 +1,62 @@
+const ParserError = require('./errors/parser-error');
+const { parseUrlVariables, getMissingProps, groupValidationErrors } = require('./utils');
+
+function validateServerVariables(parsedJSON, asyncapiYAMLorJSON, initialFormat) {
+  const srvs = parsedJSON.servers;
+  if (!srvs) return true;
+
+  const notProvidedVariables = {};
+
+  Object.keys(srvs).forEach((server) => {
+    const variables = parseUrlVariables(srvs[server].url);
+    if (!variables) return;
+
+    const missingServerVariables = getMissingProps(variables, srvs[server].variables);
+    if (missingServerVariables.length) notProvidedVariables[server] = notProvidedVariables[server] ? notProvidedVariables[server].concat(missingServerVariables) : missingServerVariables;
```

Comment:
I'm surprised Sonar is not warning you about this. Assigning properties this way is not a good practice because you can unexpectedly override "special" object properties. For instance, consider we have a variable called "constructor". This line would be doing:

```js
notProvidedVariables['constructor'] = missingServerVariables;
```

You would be changing the constructor of the object and that can cause undesired problems. See [The Dangers of Square Bracket Notation](https://github.com/nodesecurity/eslint-plugin-security/blob/master/docs/the-dangers-of-square-bracket-notation.md).

Reply:
That said, I'd use arrays and leverage Array.find which is performat enough, especially for small arrays.

Reply:
We can probably add this: https://github.com/nodesecurity/eslint-plugin-security.

Reply:
`notProvidedVariables` is an internal object of this function and doesn't have any constructors, it's purpose is to provide a structured information about not provided params/variables so later it is much easier to extract the information in  `groupValidationErrors` function. Using array here adds unnecessary complexity.

the plugin you suggested throws 
<img width="843" alt="Screenshot 2020-06-30 at 17 06 46" src="https://user-images.githubusercontent.com/6995927/86143717-12008800-baf5-11ea-8a7f-25b51367fc44.png">
and I believe it will be more annoying than useful, its fails to bring good info.

Even if someone would really provide values in AsyncAPI document that are hidden functions or something, looking at my code I don't see what could go wrong really.

Reply:
That's ok but consider it a code smell (something that in the future could potentially become problematic). Right now the variable is contained in this function and we only return true or false but a change in the future might make us think it's better to return the object instead of a boolean and suddenly it goes beyond the scope of the function.

> `notProvidedVariables` is an internal object of this function and doesn't have any constructors

Just a note: every variable in JS has a constructor.
![Screen Shot 2020-07-01 at 13 06 06](https://user-images.githubusercontent.com/242119/86237251-ba692780-bb9b-11ea-92d0-e468d384b767.png)

If you want to get rid of this "issue", consider using a Map: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map.

Anyway, not a nazi of these things myself, just consider it.

> and I believe it will be more annoying than useful, its fails to bring good info.

It's annoying because I'm doing the same thing in other places and I probably shouldn't. We can always add `// eslint-ignore` to the line.

Reply:
@fmvilas you are right, I just pushed a change and after working on refactoring I have to say I prefer working with Map than an Object really.

I would not introduce the linter though as it is still failing in many other places and I should not fix them in this pr and I'm skeptical if it comes to `ignoring`. I would prefer to have another issue for that, and if it is not in scope for 1.0 it could be marked with `good first issue` as because of the size of the changes it is a good task to get familiar with the project

Reply:
https://github.com/asyncapi/parser-js/issues/103